README.md

```diff
@@ -99,41 +136,71 @@ into the element. This is the only method that must be implemented by subclasses
-  rendered elements, for example focusing an input:
-  `this.shadowRoot.querySelector('input').focus()`.
-
-  * `renderComplete`: Returns a promise which resolves after the element next renders.
-
-  * `_requestRender`: Call to request the element to asynchronously re-render regardless
+  an optimization to avoid updating work when changes occur, which should not be rendered.
+
+  * `update()` (protected): This method calls `render()` and then uses `lit-html` to
+  render the template DOM. Override to customize how the element renders DOM. Note,
+  during `update()` setting properties does not trigger `invalidate()`, allowing
+  property values to be computed and validated.
+
+  * `render()` (protected): Implement to describe the element's DOM using `lit-html`. Ideally,
+  the `render` implementation is a pure function using only the element's current properties
+  to describe the element template. This is the only method that must be implemented by subclasses.
+  Note, since `render()` is called by `update()` setting properties does not trigger
+  `invalidate()`, allowing property values to be computed and validated.
+
+  * `finishUpdate(changedProperties): Promise?` (protected): Called after element DOM has been updated and
+  before the `updateComplete` promise is resolved. Implement to directly control rendered DOM.
+  Typically this is not needed as `lit-html` can be used in the `render` method
+  to set properties, attributes, and event listeners. However, it is sometimes useful
+  for calling methods on rendered elements, for example focusing an input:
+  `this.shadowRoot.querySelector('input').focus()`. The `changedProps` argument is an object
+  with keys for the changed properties pointing to their previous values. If this function
+  returns a `Promise`, it will be *awaited* before resolving the `updateComplete` promise.
+  Setting properties in `finishUpdate()` does trigger `invalidate()` and blocks
+  the `updateComplete` promise.
+
+  * `finishFirstUpdate(changedProperties): Promise?` (protected) Called after element DOM has been
+  updated the first time. This method can be useful for capturing references to rendered static
+  nodes that must be directly acted upon, for example in `finishUpdate`.
+
+  * `updateComplete`: Returns a promise which resolves after the element next renders.
+
+  * `invalidate`: Call to request the element to asynchronously update regardless
   of whether or not any property changes are pending.
 
+## Update Lifecycle
+
+* When the element is first connected or a property is set (e.g. `element.foo = 5`)
+  and the property's `shouldInvalidate(value, oldValue)` returns true. Then
+  * `invalidate()` tries to update the element after waiting a [microtask](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/) (at the end
+  of the event loop, before the next paint). Then
+    * `shouldUpdate(changedProps)` is called and if this returns true which it
```

Comment:
changedProps again... also below... do a search for that

Reply:
Done